tower/src/ready_cache/cache.rs

```diff
@@ -410,20 +431,40 @@ where
     type Output = Result<(K, S, CancelRx), PendingError<K, S::Error>>;
 
     fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        let mut fut = self.cancel.as_mut().expect("polled after complete");
-        if let Poll::Ready(r) = Pin::new(&mut fut).poll(cx) {
-            assert!(r.is_ok(), "cancel sender lost");
+        // Before checking whether a service ready, check to see whether
+        // readiness has been canceled.
+        let CancelRx(cancel) = self.cancel.as_mut().expect("polled after complete");
+        if cancel.canceled.load(Ordering::SeqCst) {
             let key = self.key.take().expect("polled after complete");
             return Err(PendingError::Canceled(key)).into();
         }
 
+        // If readiness hasn't been canceled, check to see whether the service
+        // is ready.
         match self
             .ready
             .as_mut()
             .expect("polled after ready")
             .poll_ready(cx)
         {
-            Poll::Pending => Poll::Pending,
+            Poll::Pending => {
+                // Before returning Pending, register interest in cancelation so
+                // that this future is polled again if the state changes.
+                let CancelRx(cancel) = self.cancel.as_mut().expect("polled after complete");
+                tokio::pin! {
+                    let cancel_notified = cancel.notify.notified();
+                }
+                let crx_ready = cancel_notified.poll(cx);
```

Comment:
i don't think this is correct. this will immediately drop the `Notified` future at the end of the `poll` call. this releases interest in the `Notify`. We would need to hold the same `Notified` future for the `Pending`'s entire lifetime, and poll the _same_ future every time, instead of creating a new one each time.

alternatively, we could replace the use of `Notify` with a lower-level primitive like [`futures::task::AtomicWaker`](https://docs.rs/futures/latest/futures/task/struct.AtomicWaker.html) -- that plus the `AtomicBool` should be sufficient here as well.